server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/zcoin/ZcoinExtension.java

```diff
@@ -0,0 +1,93 @@
+package com.generalbytes.batm.server.extensions.extra.zcoin;
+
+import com.generalbytes.batm.server.extensions.*;
+import com.generalbytes.batm.server.extensions.extra.flashcoin.sources.coinexchange.CoinexchangeRateSource;
+import com.generalbytes.batm.server.extensions.extra.zcoin.sources.FixPriceRateSource;
+import com.generalbytes.batm.server.extensions.extra.zcoin.wallets.zcoind.ZcoinRPCWallet;
+import org.graalvm.compiler.core.common.type.ArithmeticOpTable;
+
+import java.math.BigDecimal;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.StringTokenizer;
+
+public class ZcoinExtension extends AbstractExtension {
+    @Override
+    public String getName() { return "BATM Zcoin extension"; }
+
+    @Override
+    public IWallet createWallet(String walletLogin) {
+        if (walletLogin != null && !walletLogin.trim().isEmpty()) {
+            StringTokenizer st = new StringTokenizer(walletLogin, ":");
+            String walletType = st.nextToken();
+
+            if("zcoind".equalsIgnoreCase(walletType)) {
+                String protocol     = st.nextToken();
+                String username     = st.nextToken();
+                String password     = st.nextToken();
+                String hostname     = st.nextToken();
+                String port         = st.nextToken();
+                String accountName  = "";
+                if (st.hasMoreTokens()) {
+                    accountName = st.nextToken();
+                }
+
+                if (protocol != null &&
+                    username != null &&
+                    password != null &&
+                    hostname !=null &&
+                    port != null &&
+                    accountName != null) {
+                    String rpcURL = protocol + "://" + username + ":" + password
+                        + "@" + hostname + ":" + port;
+                }
+            }
+        }
+        return  null;
+    }
+
+    @Override
+    public ICryptoAddressValidator createAddressValidator(String cryptoCurrency) {
+        if (Currencies.XZC.equalsIgnoreCase(cryptoCurrency)) {
+            return new ZcoinAddressValidator();
+        }
+        return null;
+    }
+
+    @Override
+    public IRateSource createRateSource(String sourceLogin) {
+        if (sourceLogin != null && !sourceLogin.trim().isEmpty()) {
+            StringTokenizer st = new StringTokenizer(sourceLogin, ":");
+            String exchangeType = st.nextToken();
+            if ("coinmarketcap".equalsIgnoreCase(exchangeType)) {
+                String preferredFiatCurrency = Currencies.USD;
+                if (st.hasMoreTokens()) {
+                    preferredFiatCurrency = st.nextToken().toUpperCase();
+                }
+                return new CoinexchangeRateSource(preferredFiatCurrency);
```

Comment:
you are using `"coinmarketcap"` prefix for `CoinexchangeRateSource`?

Reply:
Yes, this means I am using coinmarketcap as exchange for price right? I just followed the procedure on other coins.

Reply:
No. You are instantiating `CoinexchangeRateSource`. No coinmarketcap usage in your code as far as I can see.

Reply:
Okay is there a better way to implement this? Apologies, I thought there is already an API handling this

Reply:
No problem. Just instantiate a `CoinmarketcapRateSource` instead (see for example `ElectraExtension`).